qhbmlib/inference/ebm.py

```diff
@@ -520,3 +520,145 @@ def _log_partition_forward_pass(self):
   def _sample(self, num_samples: int):
     """See base class docstring"""
     return self.distribution.sample(num_samples, seed=self.seed)
+
+
+class GibbsWithGradientsKernel(tfp.mcmc.TransitionKernel):
+  """Implements the Gibbs With Gradients update rule.
+
+  See Algorithm 1 of
+  https://arxiv.org/abs/2102.04509v2
+  for details.
```

Comment:
Let's add a blurb regarding the idea of this algorithm. For example:

> GWG adapts Gibbs sampling (a special case of Metropolis-Hastings) to more frequently consider bit-string dimensions that are more likely to require updates. For example, in sampling from a representation of the MNIST dataset one can take advantage of the fact that almost all dimensions represent the background and therefore are unlikely to change sample-to-sample.
>
>  A previously known way to achieve this behavior is to use a proposal distribution which is *locally informed* e.g. a tempered softmax over change in energy within some Hamming ball about the previous sample. This scales poorly in the data dimensionality and so the authors propose a single-evaluation first-order approximation (hence leveraging gradient information), the algorithm implemented here.

Reply:
Added an extended blurb based on your proposal